Day1

jenarp opened the pull request:
Dear Team,
I would be delighted if you could have a look at my code.
I am happy about feedback.

jenarp committed:
- I added the code to puzzle day 3 (both parts) in Day03.java.
- I wrote several test scripts (Day03Test.java).
- I resolved the comments from Michelle for Day02.java
- Raw Data is added in resources

jenarp committed:
I added the code to puzzle day 4 (both parts) in Day04.java. Additionally, I wrote several test scripts (Day04Test.java).

jenarp committed:
I added the code to puzzle day 4 (both parts) in Day04.java. Additionally, I wrote several test scripts (Day04Test.java).

jenarp committed:
Day04: I changed some modifiers.

jenarp committed:
I added the code to puzzle day 2 (both parts) in Day02.java. Additionally, I wrote several test scripts (Day02Test.java).

jenarp committed:
Adding data input to resources folder.

jenarp committed:
Delete "filereader"

jenarp committed:
Some changes for heroku

jenarp committed:
Some changes for heroku

jenarp committed:
Heroku stuff

jenarp committed:
heroku stuff

jenarp committed:
add logging

jenarp committed:
add logging

jenarp committed:
Heroku: changed method to read in data: now via "getClass().getResourceAsStream("/data/...txt")""

jenarp committed:
heroku

jenarp committed:
Changed parameter day and part in AdventOfCodeController to Integers (formerly: Strings)

jenarp committed:
Heroku: I added a default value by @RequestParam line 56 at AdventofCodeController.java

jenarp committed:
Heroku: Bug Fix in Day01 (read in data)

jenarp committed:
Heroku: AdventOfCodeServiceTest.java: modified outout to my day01 output (AoC year 2020)

jenarp committed:
Heroku: add ne FE tests fix (Michelles repo)

jenarp committed:
Heroku: add ne FE tests fix (Michelles repo)

jenarp committed:
Answer to puzzle day05: I added data in the resources folder (day05), the main and the test script (day05).

jenarp committed:
Answer to puzzle day05: Formatting of Day05.java and Day05Test.java. Change of for loop to stream() in Day05.java.

jenarp committed:
Code formatting

bonndan commented on src/main/java/org/haffson/adventofcode/days/Days.java:
```diff
@@ -38,5 +38,5 @@
      * @return the status of each part
      */
-    HashMap<String, ProblemStatusEnum> getProblemStatus();
+    Map<Integer, ProblemStatusEnum> getProblemStatus();
```
👍

bonndan commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -2,36 +2,31 @@
+    private final Map<Integer, ProblemStatusEnum> problemStatus;
+
+    // Read content of input file
+    public InputStream resource = getClass().getResourceAsStream("/data/day01/input_day01.txt");
```
Den Code würde ich ich den Konstruktor verlagern, damit er explizit aufgerufen wird. Jetzt läuft er zwar auch "da", aber implizit und ich weiß gar nicht ob vor dem Konstruktor oder danach.

Wahrscheinlich kann man dann auch auf "resource" als Feld verzichten.

bonndan commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -68,72 +59,52 @@ public String secondPart() {
-            int i = Integer.parseInt(s.next());
-            data.add(i);
+    // read raw data and transform it to String[]
+    public Integer[] getRawData(InputStream resource) {
```
Null check wäre hier gut. Muss die Methode public sein?

bonndan commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -68,72 +59,52 @@ public String secondPart() {
 
-        for (int j = 0; j < data.size(); j++) {
-            data2.add(2020 - data.get(j));
+    private int calculateNumber1(Integer[] rawData_array) {
```
Vielleicht "numbers" oä als Parameternamen verwenden...

bonndan commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -68,72 +59,52 @@ public String secondPart() {
-
+    private int calculateNumber2(Integer[] rawData_array) throws FileNotFoundException {
+        List<Integer> data = new ArrayList<>(rawData_array.length);
+        data.addAll(Arrays.asList(rawData_array));
```
Arrays.asList(rawData_array) ist schon eine Liste, da braucht man Zeile 93 nicht.

bonndan commented on src/main/java/org/haffson/adventofcode/days/day03/Day03.java:
```diff
@@ -0,0 +1,145 @@
+    private final Map<Integer, ProblemStatusEnum> problemStatus;
+
+    // Read content of input file
+    public InputStream resource = getClass().getResourceAsStream("/data/day03/input_day03.txt");
```
Mir fällt auf, dass sich viel Code gleich oder ähnlich wiederholt (stream, scanner, rawdata, problemStatus...). Man könnte vielleicht Hilfsklasse(n) bauen, die diese Funktionalität kapseln und dann von den Days genutzt werden

bonndan commented on src/main/java/org/haffson/adventofcode/days/day03/Day03.java:
```diff
@@ -0,0 +1,145 @@
+    public InputStream resource = getClass().getResourceAsStream("/data/day03/input_day03.txt");
+    private final String[] data = getRawDataAsArray(resource);
+
+    @Autowired
```
Autowired benötigt man nicht mehr bei Spring Beans (@Component, @Service, @NeMengeMehr). Und hier hat der Konstruktor ja auch gar keine Parameter/Dependencies, die Spring autowiren könnte.

bonndan commented on src/main/java/org/haffson/adventofcode/days/day04/Day04.java:
```diff
@@ -0,0 +1,170 @@
+    public int getNumberValidPassports(String[] rawData_array) {
+        int numberOfValidPassports = 0;
+        for (String passport : rawData_array) {
+            Set<String> validKeys = new HashSet<>(Arrays.asList("byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"));
```
validKeys sind ja immer gleich. Es reicht die einmal zu initialisieren (z.B. als Feld bzw. sogar statisches Feld).

bonndan commented on src/main/java/org/haffson/adventofcode/service/AdventOfCodeService.java:
```diff
@@ -43,13 +43,18 @@ public AdventOfCodeService(List<Days> daysSolutions) {
         if (!isProblemSolvedForPart(thisDaysClass, part)) {
             throw new PuzzleNotSolvedYetException(new Throwable());
-        } else if (("1").equals(part)) {
+        } else if ((part).equals(1)) {
```
Alt+Enter : "remove unnecessary parentheses"

Alt+Enter : "flip equals" -> momentan ist das nicht null-safe, falls part null ist.

jenarp replied:
Ich habe in der Methode folgendes hinzugefügt:
Objects.requireNonNull(day, "day is null");
Objects.requireNonNull(part, "part is null");

Zusätzlich funktioniert equals() nicht mit int. Und 1 und 2 sind ja int, weswegen ein flip nur möglich wäre, wenn ich diese zu Integers umwandel. Sollte man das tun, oder es dann so lassen? Auch "==" würde ja jetzt funktionieren, benutzt man dann lieber == oder equals()?

MatthiasMolitor replied:
Bei primitiven Typen wie `ìnt` lieber `==`.

Ein bisschen aufpassen musst du, wenn einer der Operanden doch ein Objekt ist (zB. `Integer`): Dann führt die Belegung mit `null` zu einer Exception:

    Integer a = null;
    var eq = a == 1; // Exception

Mit `Objects.equals()` muss man sich über solche Details keine Gedanken machen:

    Integer a = null;
    var eq = Objects.equals(a, 1);

bonndan reviewed (changes requested):
Ich habe erstmal  nur ein paar grundlegende Dinge kommentiert und auch nichts, was doppelt auftritt.

MatthiasMolitor commented on frontend/package.json:
```diff
@@ -10,7 +10,7 @@
     "react": "^16.12.0",
     "react-dom": "^16.12.0",
-    "react-native": "^0.61.5",
+    "react-native": "^0.65.1",
```
Hatte die neuere Version einen Bug?

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -2,36 +2,31 @@
-        this.problemStatus.put("1", ProblemStatusEnum.SOLVED);
-        this.problemStatus.put("2", ProblemStatusEnum.SOLVED);
+        this.problemStatus.put(1, ProblemStatusEnum.SOLVED);
+        this.problemStatus.put(2, ProblemStatusEnum.SOLVED);
```
Alternative, wenn man die Map anschließend nicht mehr ändern möchte:

    Map.of(
       1, ProblemStatusEnum.SOLVED,
       2, ProblemStatusEnum.SOLVED
    );

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -40,23 +35,19 @@ public int getDay() {
-            return "Product 2: " + calculateNumber2();
+            return "Product 2: " + calculateNumber2(data);
         } catch (FileNotFoundException e) {
             return "error";
```
Die Exception wird hier ignoriert, wodurch die Fehlerinformationen verloren sind.

Die möchte man eigentlich immer mindestens loggen:

    log.error("{}", e.getMessage(), e);

... oder gar nicht behandeln, wenn das nicht sinnvoll möglich ist.
Dazu entweder gar nicht abfangen oder verpacken und weiter werfen:

    throw new RuntimeException("Oh... that is... unfortunate.", e);

jenarp replied:
Dass der InputStream null ist, wird nun an einer anderen Stelle abgefangen, ich habe eine Hilfsklasse "DataLoader" geschrieben, in der für jeden Tag die Methoden stehen, die die Daten einlesen. Dort steht nun so was:
InputStream dataIn = Objects.requireNonNull(DataLoader.class.getResourceAsStream("/data"
                        + CheckStringisEmpty.requireNonNullAndNonEmpty(path)),
                "Data InputStream must not be null.");

MatthiasMolitor replied:
Kann die `FileNotFoundException` dann entfallen?

jenarp replied:
Ich würde sagen, ja...

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -68,72 +59,52 @@ public String secondPart() {
         }
-        s.close();
+        Integer[] rawData_array = new Integer[rawData.size()];
+        for (int i = 0; i < rawData.size(); i++) rawData_array[i] = Integer.parseInt(rawData.get(i));
```
Alternativ kann man das auch in einem    Stream machen:

    Integer[] rawData_array = rawData.stream()
        .map(Integer::parseInt)
        .toArray(Integer[]::new);

Wenn du nicht unbedingt ein Array brauchst, dann kannst du in deinem Code auch direkt mit einer Liste (`List<Integer>`) arbeiten. Listen sind oft einfacher zu handhaben.

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day02/Day02.java:
```diff
@@ -87,57 +88,63 @@ public String secondPart() {
-    public List<Data> getData(List<String> rawData) {
+    public static String requireNonNullAndNonEmpty(String string) {
+        if (StringUtils.isEmpty(string)) {
+            throw new NullPointerException("The string is null or empty");
```
Hier wäre wahrscheinlich eine `IllegalArgumentException` noch besser.

Bei einer `NullPointerException` würde ich am ehesten einen fehlerhaften Zugriff oder einen anderen groben Fehler vermuten.

    String test = null;
    int length = test.length();

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day02/Day02.java:
```diff
@@ -87,57 +88,63 @@ public String secondPart() {
-
-            // match patterns to determine min and max number of the letter in pwd, the searched letter and pwd itself
+            // match patterns to determine min and max number of the letter in password,
+            // the searched letter and password itself
```
:+1:

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day03/Day03.java:
```diff
@@ -0,0 +1,145 @@
+                numTrees++;
+            }
+        }
+        return "" + numTrees;
```
Hier könntest du auch `Integer.toString(numTrees);` benutzen.
Oder direkt einen `int` zurückgeben.

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day04/Day04.java:
```diff
@@ -0,0 +1,170 @@
+        for (String passport : rawData_array) {
+            int count = 0;
+            Map<String, String> passports = new HashMap<>();
+            Set<String> validKeys = new HashSet<>(Arrays.asList("byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"));
```
Eine Alternative zu `Arrays.asList()` ist `List.of()`.

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day04/Day04.java:
```diff
@@ -0,0 +1,170 @@
+                    count++;
+                }
+                // check body height
+                String hgt = passports.get("hgt");
```
Wenn du die Variable direkt `height` nennst, dann hat der Leser es etwas einfacher.

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day05/Day05.java:
```diff
@@ -0,0 +1,124 @@
+    // seatID is row*8 + column
+    public List<Integer> getSeatID(@NonNull final InputStream dataIn) {
+        return getRawdataAsList(Objects.requireNonNull(dataIn, "Data Inputstream is null.")).stream()
+                .map(ss -> getRowOrCol(ss.substring(0, 7), 128, 'F', 'B') * 8
```
Wofür steht in diesem Kontext `ss`? Würde ich eindeutig benennen.

MatthiasMolitor commented on src/main/java/org/haffson/adventofcode/days/day03/Day03.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    public String[] getTestData() {
+        final String test = "..##.......\n" +
+                "#...#...#..\n" +
+                ".#....#..#.\n" +
+                "..#.#...#.#\n" +
+                ".#...##..#.\n" +
+                "..#.##.....\n" +
+                ".#.#.#....#\n" +
+                ".#........#\n" +
+                "#.##...#...\n" +
+                "#...##....#\n" +
+                ".#..#...#.#";
```
Wenn du schon Java 16 benutzt, dann kannst du auch die neuen Multiline-Strings benutzen:

    final String test = """
        ..##.......
        #...#...#..
        [...]
    """;

jenarp committed:
day01, day02, day03, day04: using utility class to read in raw Data, format scripts, change modifiers to private and final whenever possible.

jenarp committed:
day01, day02, day03, day04, day05: using utility class to read in raw Data (DataLoader.java), to create problemStatus Maps (Problemstatus.java) and to throw illegal Argument Exception if Strings are empty (CheckStringIsEmpty.java).
 Additionally, some script formatting, modifiers changing to private and final whenever possible and variable renaming.

bonndan commented on src/main/java/org/haffson/adventofcode/ProblemStatusEnum.java:
```diff
@@ -6,12 +6,18 @@
  * @author Michelle Fernandez Bieber
  */
 public enum ProblemStatusEnum {
```
Den Ansatz finde ich gut 👍

bonndan commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -1,32 +1,33 @@
-    private final Integer[] data = getRawData(resource);
+    public Day01(String filename) {
+        //get data
+        this.numbers = DataLoader.getDataDay01("/day01/" + filename, "\n");
```
Gut, dass das Laden nun extern passiert. Hier wird DataLoader statisch genutzt, was den Nachteil hat, dass man ihn im Test immer mittestet und sich so von ihm und (wahrscheinlich) dem Dateisystem abhängig macht.

jenarp replied:
Ja,
das stimmt, aber ich wollte dynamisch bleiben bzgl der Dateien, die ich laden möchte, denn ich habe pro Rätsel oftmals mehr als einen Dateiinput. Manchmal bis zu 4 (z.b. data part 1, data part 2, test data part 1, test data part 2). Ich habe lange überlegt, wie ich es am besten mache und habe mich für diese Form entschieden.

jenarp replied:
Gibt es eine bessere Art mit verschiedenen Datensätzen umzugehen?

MatthiasMolitor replied:
Du könntest den `DataLoader` als Instanz an den Konstruktor übergeben und  die `getData`-Methode nicht statisch machen:

    public Day01(DataLoader dataLoader, String filename) {
        //get data
        this.numbers = dataLoader.getDataDay01("/day01/" + filename, "\n");
    }

Dadurch kannst du den `DataLoader` im Test austauschen und durch eine Komponente ohne Dateisystemzugriff austauschen (z.B. ein Mock-Object).

Nächster Level: Anstatt einen `DataLoader` zu übergeben definierst du ein Interface, das `DataLoader` implementiert wird:

    public interface Loader {
        InputStream load(String path);
    }

    Day1(Loader dataLoader) {
    }

Wenn du das Interface als functional definierst, dann kannst du Lambdas anstelle von konkreten Instanzen benutzen:

    @FunctionalInterface
    public interface Loader {
        // ...
    }

Damit brauchst du in Tests nicht einmal Mock-Objekte:

    final var day1 = new Day((path) -> {
        // Hier Input-Stream erstellen, z.B. in-memory aus String.
        return input;
    });

bonndan commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -1,32 +1,33 @@
+        //get data
+        this.numbers = DataLoader.getDataDay01("/day01/" + filename, "\n");
+        // set ProblemStatus
+        this.problemStatus = ProblemStatus.getProblemStatusMap(1, 2,
```
Hier ist das vielleicht erstmal egal, aber es ist ja so, dass man im Konstruktor schon weiß, dass beide Aufgaben gelöst sind. Man würde das in anderen Domänen/Projekten/etc. vielleicht eher so machen, dass am Ende eine Aufgabe der Status gesetzt wird.

jpgoelz replied:
Keine schlechte Idee. 🤔

bonndan commented on src/main/java/org/haffson/adventofcode/days/day04/Day04.java:
```diff
@@ -111,26 +87,26 @@ public int getRestrictedNumberValidPassports(String[] rawData_array) {
                 // check birth year
-                int byr = Integer.parseInt(passports.get("byr"));
-                if (1920 <= byr && byr <= 2002) {
+                int birthYear = Integer.parseInt(passports.get("byr"));
```
👍 für die Umbenennung

bonndan commented on src/main/java/org/haffson/adventofcode/days/day05/Day05.java:
```diff
@@ -1,107 +1,83 @@
-        return getRawdataAsList(Objects.requireNonNull(dataIn, "Data Inputstream is null.")).stream()
-                .map(ss -> getRowOrCol(ss.substring(0, 7), 128, 'F', 'B') * 8
-                        + getRowOrCol(ss.substring(7), 8, 'L', 'R'))
+    private List<Integer> getSeatID(@NonNull List<String> boardingPasses) {
```
Ich weiß nicht, ob ich @NonNull an private Methoden schreiben würde. @Matthimatiker @basseur Meinung pls.

Und "final" ist weg, warum?

MatthiasMolitor replied:
Wenn ich weiß, ob eine Variable `null` sein kann/darf (oder eben nicht), dann würde ich es immer annotieren.

bonndan commented on src/main/java/org/haffson/adventofcode/utils/CheckStringisEmpty.java:
```diff
@@ -0,0 +1,20 @@
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Component;
+
+@Component
```
Das muss keine Component sein, weil es nur eine statische Methode gibt, die man nutzt.

bonndan commented on src/main/java/org/haffson/adventofcode/utils/CheckStringisEmpty.java:
```diff
@@ -0,0 +1,20 @@
+    public static String requireNonNullAndNonEmpty(String string) {
+        if (StringUtils.isEmpty(string)) {
+            throw new IllegalArgumentException("The string is null or empty");
+        } else {
```
Den meisten Code kann man ohne "else" schreiben. Hier auch.

bonndan commented on src/main/java/org/haffson/adventofcode/utils/CheckStringisEmpty.java:
```diff
@@ -0,0 +1,20 @@
+
+    }
+
+    public static String requireNonNullAndNonEmpty(String string) {
```
Die Hilfsmethode finde ich erstmal gut. Man muss in so einem Fall allerdings

- gucken, was es ist libraries schon gibt 
- tests dafür schreiben
- dokumentieren

bonndan commented on src/main/java/org/haffson/adventofcode/utils/DataLoader.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Component
```
Entweder statische Nutzung oder @Component (s.o.)

bonndan commented on src/main/java/org/haffson/adventofcode/utils/DataLoader.java:
```diff
@@ -0,0 +1,56 @@
+    private DataLoader() {
+    }
+
+    public static List<String> getRawDataAsList(@NonNull String path, @NonNull String delimiter) {
```
Finde ich gut 👍

jenarp committed:
Create mock Objects of DataLoader for all testing scripts

jenarp committed:
Test script CheckStringIsEmpty.java. Test boundary condition, i.e. (String is null, empty, existing...)

jenarp committed:
Added Documentation to CheckStringIsEmpty.java

jenarp committed:
Split: test data is only used in Test scripts and input Data only for the application.
-> day's constructors do not need "String filename" as argument anymore, as data is loaded in DataLoader from now on.

jpgoelz commented on src/main/java/org/haffson/adventofcode/utils/CheckStringIsEmpty.java:
```diff
@@ -0,0 +1,21 @@
+    private CheckStringIsEmpty() {
+    }
+
+    /*
+    requireNonNullAndNonEmpty() checks if strings are null or empty
+    and if true throws IllegalArgumentException.
+    It helps crashing code "fail-fast" before code has already performed some side effects.
+     */
```
Für Dokumentationskommentare solltest du JavaDocs verwenden. Die erhältst du, indem du `/**` eintippst und Enter drückst. Hast du an anderen Stellen ja schon verwendet.

jpgoelz commented on src/main/java/org/haffson/adventofcode/utils/CheckStringIsEmpty.java:
```diff
@@ -0,0 +1,21 @@
+     */
+    public static String requireNonNullAndNonEmpty(String string) {
+        if (StringUtils.isEmpty(string)) {
+            throw new IllegalArgumentException("The string is null or empty: " + string);
```
Macht es Sinn, einen leeren String der Message anzufügen? 🤔

jpgoelz commented on src/main/java/org/haffson/adventofcode/utils/DataLoader.java:
```diff
@@ -0,0 +1,56 @@
+    public DataLoader() {
+    }
+
+    private List<String> getRawDataAsList(@NonNull String path, @NonNull String delimiter) {
```
Ich finde, es macht für die Lesbarkeit Sinn, die requireNonNullAndNonEmpty-Checks auf die Parameter der Methode direkt am Anfang zu machen. Dadurch wird der restliche Code deutlich besser lesbar.

Viel wichtiger aber: Es Sinn, eine Methode so früh wie möglich aussteigen zu lassen, wenn der Input zu Fehlern führen wird. Wenn du schon am Anfang weißt, dass `delimiter` kein gültiges Argument ist, musst du den Input ja gar nicht erst in `dataIn` einlesen und mit dem Scannen anfangen, nur um dann im `while` erst eine Exception zu erhalten.

jpgoelz commented on src/main/java/org/haffson/adventofcode/utils/DataLoader.java:
```diff
@@ -0,0 +1,56 @@
+        return rawData;
+    }
+
+    public List<Integer> getDataDay01() {
+        return getRawDataAsList("/day01/input_day01.txt", "\n").stream()
+                .map(Integer::parseInt)
+                .collect(Collectors.toList());
+    }
+
+    public List<String> getDataDay02() {
+        return getRawDataAsList("/day02/input_day02.txt", "\n");
+    }
+
+    public List<String> getDataDay03() {
+        return getRawDataAsList("/day03/input_day03.txt", "\n");
+    }
+
+    public List<String> getDataDay04() {
+        return getRawDataAsList("/day04/input_day04.txt", "\n\n");
+    }
+
+    public List<String> getDataDay05() {
+        return getRawDataAsList("/day05/input_day05.txt", "\n");
+    }
+
+    public List<String> getDataDay06() {
+        return getRawDataAsList("/day06/input_day06.txt", "\n");
+    }
```
Ich würde eigentlich nicht für jeden Tag eine eigene Methode machen, das wird schnell unübersichtlich und enthält mir hier zu sehr anwendungsspezifische Details in einer eigentlich allgemeinen Klasse.

jenarp replied:
Ja, ich hatte das auch mal viel allgemeiner. Ich wollte anfangs das Dateneinlesen sehr flexibel gestalte, da ich nicht nur die Input Daten, sondern auch z.B. Testdaten für die einzelnen Tage einlesen wollte. 
Jetzt habe ich es aber so geändert, (mit Absprache mit Daniel) dass die Testdaten alle händisch in ein Mock Objekt übergeben werden und ich so Unit Tests auf die Days machen kann, ohne den DataLoader zu instanzieren.
Jetzt ist die Aufgabe vom DataLoader nur die Daten für die einzelnen Tage bereitzustellen. Jetzt kann man einzeln und unabhängig voneinander die Days und den DataLoader testen.

jpgoelz replied:
Klingt gut! 🎉

jpgoelz commented on src/main/java/org/haffson/adventofcode/utils/ProblemStatus.java:
```diff
@@ -0,0 +1,21 @@
+    private ProblemStatus() {
+    }
+
+    public static Map<Integer, ProblemStatusEnum> getProblemStatusMap(int part1, int part2,
+                                                                      ProblemStatusEnum status1, ProblemStatusEnum status2) {
```
Ich nutze gerne folgenden Stil, weil er etwas übersichtlicher ist. Ist aber natürlich geschmacksache.
```suggestion
    public static Map<Integer, ProblemStatusEnum> getProblemStatusMap(
        int part1,
        int part2,
        ProblemStatusEnum status1,
        ProblemStatusEnum status2
    ) {
```

jpgoelz commented on src/main/java/org/haffson/adventofcode/utils/ProblemStatus.java:
```diff
@@ -0,0 +1,21 @@
+
+    public static Map<Integer, ProblemStatusEnum> getProblemStatusMap(int part1, int part2,
+                                                                      ProblemStatusEnum status1, ProblemStatusEnum status2) {
+        final Map<Integer, ProblemStatusEnum> problemStatus = new HashMap<>();
+
+        problemStatus.put(part1, status1);
+        problemStatus.put(part2, status2);
+        return problemStatus;
```
Matthias hatte das glaube ich schon mal angemerkt. Du möchtest den Problem-Status ja nicht mehr im laufenden Programm endern, weshalb deine Map unmodifiable sein kann. Da kannst du einfach folgende Methode nutzen und direkt returnen:
```suggestion
        return Map.of(
            part1, status1,
            part2, status2
            );
```

jenarp replied:
Hallo Jan,
danke für Deine hilfreichen Kommentare. Leider gibt bei mir Map.of immer "Cannot resolve method". Ich glaube, ich benutze bei AoC Java 8 (ich bin mir aber nicht sicher). Und bei Java 8 gibt es (ich bin mir wieder nicht sicher :-D) kein Map.of...

jpgoelz replied:
Tatsächlich! Das kannst du in deiner `build.gradle` sehen. Dort steht

```
sourceCompatibility = 1.8
targetCompatibility = 1.8
```

Diesen Monat ist sogar schon Java 17 herausgekommen. Als optionale Aufgabe kannst du nach Abschluss dieses PR vielleicht ein Update vornehmen. Die Schritte wären:
* Version in der `build.gradle` erhöhen
* Version in den Project Settings von IntelliJ anpassen (`Strg`+`Shift`+`Alt`+`S`)
* Dein Code wird dann eventuell ein paar Fehler haben, die aber relativ leicht zu beheben sein sollten.

Aber wie gesagt: optional. Aber dann lernst du wenigstens auf dem aktuellen Stand. 👍

jenarp replied:
Sehr cool! Ja, das werde ich gerne machen. Hatte auch schon danach geguckt, wie man das am besten anstellt. Danke für die Anleitung.

jpgoelz replied:
Keine Gewähr für Vollständigkeit. 😅

jpgoelz commented on src/main/java/org/haffson/adventofcode/days/day01/Day01.java:
```diff
@@ -40,100 +39,59 @@ public int getDay() {
+    }
+
+
+    // utility method: subtract numbers by 2020
```
JavaDoc. Außerdem: Du substractest nicht by, sondern from.

```suggestion
    /**
     * Utility method that subtracts each number from 2020.
     */
```

jpgoelz reviewed (commented):
Ich habe mal einen Teil gereviewt, aber nicht alles.

jpgoelz commented:
Noch etwas, das mit gerade in den Sinn kommt: Vielleicht arbeitest du die Kommentare hier mal ab, mergst den Pullrequest und stellst dann in Zukunft einzelne Pullrequests für thematisch zusammenhängende Themen, wie z.B. einen neu hinzugefügten Tag.

Denn dein PR für "Day1" geht schon bis Tag 5. 😉
